### ISSUE TITLE
CMake: compile using same compilers defined in Homebrew formula

### ISSUE THREAD

**Issue by dmey** [WRF-CMake/homebrew-wrf#1]
When using Homebrew to compile on systems with multiple compilers CMake will pick the first compiler in the PATH. Instead, we should enforce CMake to use the same compilers as defined in the in the Homebrew formula.

**Comment by zbeekman** [comment]
Yes and no. Homebrew sets up a sanitized build environment. Currently, the only case that will cause this to happen is if you put another compiler into `/usr/local` AFAIK.

Did this happen to you? It's happened to me before by installing the intel compiler into the cellar `/usr/local/Cellar/intel-parallel-studio/<version>` and then `brew link intel-parallel-studio`. But I don't *think* your path will stay the same in the sanitized environment.

Yup I just confirmed this is true. You can too. e.g.,

```
export PATH=/foo/bar/ifort:$PATH
brew install --interactive --build-from-source opencoarrays
# wait for the download to happen and to be dropped into the "super-env"
echo $PATH
# output: /usr/local/Cellar/modules/4.2.3/bin:/usr/local/Homebrew/Library/Homebrew/shims/mac/super:/usr/local/opt/cmake/bin:/usr/local/opt/gcc/bin:/usr/local/opt/openssl/bin:/usr/local/opt/libevent/bin:/usr/local/opt/open-mpi/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/opt/fzf/bin
```

**Comment by zbeekman** [comment]
TLDR; only stuff needed by the formula, standard system defaults, and homebrew compiler stubs will be in your environment. In an interactive session you may get some stuff from your dot files, but that won't be there during an actual non-interactive build.

EDIT: wrong wording. `s/ships/stubs/` no idea where I got ships from...

**Comment by dmey** [comment]
To clarify there are no issue with the C compiler, the issue are with the Fortran compiler on my machine. I have `ifort` installed under `/opt/intel/compilers_and_libraries_2019.1.144/mac/bin/intel64/ifort`. This is Intel's default for a system install. This path is also added before others so other users that may use the same installation method may end up with the same issue. Passing `gfortran` to CMake solves the issue but as we discussed this may not be a good solution. Passing a the fortran compiler flag to Homebrew would be an alternative but I could not find it from https://docs.brew.sh/Manpage. I do not think that we should handle this issue by documentation alone.

**Comment by zbeekman** [comment]
Agreed. So CMake is finding and using ifort from `/opt/intel` when you try
to use the Homebrew formula?

Do you have `FC` or `F77` and friends set in your environment?

If not this is really something I need to fix on the Homebrew side. As a
temporary work around we can force the issue in the formula.

Once I hear back RE `FC` & friends I’ll push a commit.

**Comment by dmey** [comment]
> So CMake is finding and using ifort from `/opt/intel` when you try
to use the Homebrew formula?

Yes

> Do you have `FC` or `F77` and friends set in your environment?

No, in fact, when I want to compile something with GCC I normally pass those flags before the CMake command e.g. `CC=gcc FC=gfortran cmake ..`

> If not this is really something I need to fix on the Homebrew side. As a
temporary work around we can force the issue in the formula.

Agree, so are you happy to set the `CMAKE_Fortran_COMPILER` to `gcc` as a temp workaround for now?

> Once I hear back RE `FC` & friends I’ll push a commit.

Sounds good -- Thanks!

**Comment by zbeekman** [comment]
I think I have a slightly cleaner approach for setting the Fortran compiler
in the formula via FC env var. I’ll push it in a few minutes and post
instructions so you can test the code in the PR.

**Comment by dmey** [comment]
Sure, thanks.

**Comment by zbeekman** [comment]
@dmey can you please do me a favor and paste the output of `brew --env` and `brew config` here?

Also, I've pushed my changes that should resolve your ifort issue. If you wouldn't mind, could you please test drive them to ensure that the correct Fortran compiler is being picked up? To test the PR please do `brew install -v "https://raw.githubusercontent.com/WRF-CMake/homebrew-wrf/d3f6cf8d87479b7cbc0aab27b5747bb73b188788/wrf.rb"`

Thanks!

**Comment by dmey** [comment]
> @dmey can you please do me a favor and paste the output of brew --env and brew config here?

- `brew --env`
```
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
MAKEFLAGS: -j4
CMAKE_PREFIX_PATH: /usr/local
CMAKE_INCLUDE_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/libxml2:/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.14
HOMEBREW_GIT: git
HOMEBREW_SDKROOT: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
ACLOCAL_PATH: /usr/local/share/aclocal
PATH: /usr/local/Homebrew/Library/Homebrew/shims/mac/super:/usr/bin:/bin:/usr/sbin:/sbin
```

- `brew config`
```
HOMEBREW_VERSION: 2.1.4-77-gdb58b9f
ORIGIN: https://github.com/Homebrew/brew
HEAD: db58b9f41b70a331dbe9b8371527a23e8ddcc718
Last commit: 7 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 29869377376d6efaab01a0cce99f0fd756b9bd32
Core tap last commit: 36 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_NO_ANALYTICS_THIS_RUN: 1
CPU: quad-core 64-bit haswell
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1001
Git: 2.20.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 11.0.1
macOS: 10.14.5-x86_64
CLT: 10.2.1.0.1.1554506761
Xcode: N/A
CLT headers: 10.2.1.0.1.1554506761
XQuartz: 2.7.11 => /opt/X11
```

> Also, I've pushed my changes that should resolve your ifort issue. If you wouldn't mind, could you please test drive them to ensure that the correct Fortran compiler is being picked up? To test the PR please do brew install -v "https://raw.githubusercontent.com/WRF-CMake/homebrew-wrf/d3f6cf8d87479b7cbc0aab27b5747bb73b188788/wrf.rb"

👍 Thank you, I can confirm that it is now picking up `gfortran` instead of `ifort`. The formula builds correctly.

**Comment by dmey** [comment]
@zbeekman many thanks for this -- the formula now builds correctly for both WPS and WRF.

@letmaik one thing I am not so sure about is the install path. At the moment we have `/usr/local/Cellar/wrf/` as the root and then we version it like `4.1.wrf-cmake`.
I would prefer to keep the same versioning as we do for the releases so this be `4.1.0` without the suffix `wrf-cmake`. Additionally, WPS is installed under  4.1.wrf-cmake like `4.1.wrf-cmake/WPS`. This should really be at the same level as wrf I think.

As an alternative folder naming/arrangement we could have"

```
- wrf-cmake/  # Here we can also use the term `arw` instead
   - wrf/
      - 4.1.0/
   - wps/
      - 4.1.0/
```

**Comment by zbeekman** [comment]
> ```
> HOMEBREW_CC: clang
> HOMEBREW_CXX: clang++
> MAKEFLAGS: -j4
> CMAKE_PREFIX_PATH: /usr/local
> CMAKE_INCLUDE_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/libxml2:/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
> CMAKE_LIBRARY_PATH: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
> PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.14
> HOMEBREW_GIT: git
> HOMEBREW_SDKROOT: /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
> ACLOCAL_PATH: /usr/local/share/aclocal
> PATH: /usr/local/Homebrew/Library/Homebrew/shims/mac/super:/usr/bin:/bin:/usr/sbin:/sbin
> ```

Wow, CMake *REALLY* wants to find `ifort`. It must explicitly go looking for it even when it's not on your path. I wonder if adding shims for GFortran will fix this or if we (homebrew) will need to take even more drastic measures. Crazy.

**Comment by dmey** [comment]
Indeed -- I am surprised that this issue hasn't come up already though? I'd have guessed that `ifort` was popular!

**Comment by zbeekman** [comment]
> At the moment we have `/usr/local/Cellar/wrf/` as the root and then we version it like `4.1.wrf-cmake`.

This directory structure is the homebrew standard convention. If the formula then follows standard coding conventions and installs executables into the `bin` subdirectory, libraries into the `lib` subdirectory, etc. those things will get sym-linked into `/usr/local/{bin,lib}` etc. Wacky directory structures do not get automatically linked. You can provide formula with canonical install directories and prevent linkage (usually if there's a good reason, e.g., already provided by macOS) by specifying the formula as `:keg_only`.

I set the "stable" version to be (pinned to) the current last commit on the `wrf-cmake` branch. Because it's not a tag, or a normal github tarball, Homebrew has no means of inferring a version from the "stable" download location info, and it's being explicitly set: https://github.com/WRF-CMake/homebrew-wrf/pull/4/files#diff-843fc2a21e7e4a83695377ee56405da5R10 You can change this to whatever you like.

Once you tag a new release, you should probably go back to the normal `url` + `sha256` stable download source. At this point, unless you explicitly override it via `version`, the version will be extracted from the tarball name, which will be set by how you tag the repository.

> Additionally, WPS is installed under 4.1.wrf-cmake like `4.1.wrf-cmake/WPS`. This should really be at the same level as wrf I think.

It sounds like WRF needs access to the WPS build directories. WPS could be specified as a separate formula, and the build directories could be installed. However, a cleaner approach would probably be to setup a CMake package config file for WPS that gets installed on the default CMake search path, and use `find_package(WPS)` from within WRF's build system.

Alternatively, it wouldn't be too hard to adjust the build system to nearly match @dmey's suggestion above:

```
/usr/local/wrf-cmake/<version>/wrf
/usr/local/wrf-cmake/<version>/wps
```

**Comment by letmaik** [comment]
@dmey I'd like to avoid touching the build system too much at this point. @zbeekman's suggestion would be easy and can be done in the formula alone. We would install WRF in the `wrf` subfolder, and WPS in the `wps` subfolder (as is done already for WPS, though it's uppercase currently). Would that work? I'm not a big fan of linking it into `/usr/local` as such a structure should rather belong into `/opt`. Would keeping it only in the cellar be fine for our purposes?

**Comment by letmaik** [comment]
@zbeekman Let's rename the formula to `wrf-cmake` to enable your structure and also provide some separation.

**Comment by zbeekman** [comment]
> @zbeekman Let's rename the formula to `wrf-cmake` to enable your structure and also provide some separation.

Done in https://github.com/WRF-CMake/homebrew-wrf/pull/4/commits/64e0cb23b8a896141ffe1724d5e38fac32a753b6 and https://github.com/WRF-CMake/homebrew-wrf/pull/4/commits/ee98f3d69ff519ad4bdb931f0cef1d3114d021d7
